include/libsemigroups/schreier-sims-helper.hpp

```diff
@@ -0,0 +1,182 @@
+//
+
+// This file contains an implementation of a backtrack search for computing
+// the intersection of two permutations given by Schreier-Sims algorithm.
+
+// TODO:
+// 1. Implement orbit refinement heuristic for intersection.
+// 2. Make the Screier-Sims object during runtime, since we compute the
+//    stabilizers of the intersection already.
+// 3. Refactor for more generality.
+
+#ifndef LIBSEMIGROUPS_SCHREIER_SIMS_HELPER_HPP_
+#define LIBSEMIGROUPS_SCHREIER_SIMS_HELPER_HPP_
+
+#include <cstddef>  // for size_t
+
+#include "containers.hpp"     // for Array2
+#include "schreier-sims.hpp"  // for SchreierSims, SchreierSims<>::ele...
+#include "transf.hpp"         // for LeastPerm
+#include "types.hpp"          // for SmallestInteger
+
+#include <iostream>
+
+namespace libsemigroups {
+
+  namespace schreier_sims_helper {
+
+    template <size_t N>
+    using point_type = typename SchreierSims<N>::point_type;
+
+    template <size_t N>
+    using element_type = typename SchreierSims<N>::element_type;
+
+    // This might not be correct for general traits
+    template <size_t N>
+    using One = typename SchreierSims<N>::One;
+
+    // This might not be correct for general traits
+    template <size_t N>
+    using Product = typename SchreierSims<N>::Product;
+
+    //! Find the intersection of two permutation groups.
+    //!
+    //! Modifies the first parameter \p T to be the Schreier-Sims object
+    //! corresponding to the intersection of \p S1 and \p S2.
+    //!
+    //! \tparam N the largest point not fixed by the permutations in the
+    //! permutation groups.
+    //!
+    //! \param T the Screier-Sims object holding the result.
+    //! \param S1 the first semigroup of the intersection.
+    //! \param S2 the second group of the intersection.
+    //!
+    // TODO(later) example
+    template <size_t N>
+    void intersection(SchreierSims<N>& T,
+                      SchreierSims<N>& S1,
+                      SchreierSims<N>& S2) {
+      S1.run();
+      S2.run();
+      if (S2.base_size() < S1.base_size()) {
+        intersection(T, S2, S1);
+        return;
+      }
+
+      // If N <= 1 then both S1, S2 are trivial.
+      if (N <= 1) {
+        T.run();
+        return;
+      }
+
+      // Note that if N-1 points are fixed then the N-th point is also fixed.
+      // So if base contains all N points, then we lose nothing by discarding
+      // the last point in the base.
+      size_t base_size = S1.base_size();
+      if (base_size == N)
+        base_size = N - 1;
+
+      SchreierSims<N> S2B;
+      for (size_t depth = 0; depth < base_size; ++depth) {
+        S2B.add_base_point(S1.base(depth));
+      }
+      for (size_t i = 0; i < S2.number_of_generators(); ++i) {
+        S2B.add_generator(S2.generator(i));
+      }
+      S2B.run();
+
+      for (size_t depth = 0; depth < base_size; ++depth) {
+        LIBSEMIGROUPS_ASSERT(S1.base(depth) == S2B.base(depth));
+      }
+
```

Comment:
```suggestion
#endif
```

Reply:
Or you could delete this loop altogether, it looks like you check for the same thing lower down.